File: move_group.py
# save as: scene_sync_cube.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
import tf2_ros
from tf2_ros import TransformException
from moveit_commander import MoveGroupCommander, PlanningSceneInterface
from shape_msgs.msg import SolidPrimitive

CUBE_FRAME   = "_09_gelatin_box"   # <-- set to your Isaac child frame
WORLD        = "world"
OBJECT_NAME  = "isaac_box"
BOX_SIZE     = (0.20, 0.15, 0.10)  # meters (x, y, z)

FLOOR_NAME   = "floor"
FLOOR_SIZE   = (5.0, 5.0, 0.02)    # big slab as ground
FLOOR_Z      = -0.01               # top flush at z=0

class SceneSync(Node):
    def __init__(self):
        super().__init__("scene_sync_cube")
        self.tf_buffer = tf2_ros.Buffer(cache_time=rclpy.duration.Duration(seconds=5.0))
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)
        self.scene = PlanningSceneInterface(synchronous=True)

        # one-shot timer to add ground after PlanningSceneMonitor is up
        self._floor_added = False
        self.create_timer(1.0, self._add_floor_once)

        # periodic update of Isaac cube into planning scene
        self.create_timer(0.1, self._tick)  # 10 Hz

    def _add_floor_once(self):
        if self._floor_added:
            return
        # if PlanningScene not ready yet, names may be empty; retry next tick
        names = set(self.scene.get_known_object_names())
        if FLOOR_NAME not in names:
            pose = PoseStamped()
            pose.header.frame_id = WORLD
            pose.pose.position.z = FLOOR_Z
            pose.pose.orientation.w = 1.0
            self.scene.add_box(FLOOR_NAME, pose, size=FLOOR_SIZE)
            self.get_logger().info("Added ground plane to planning scene.")
        self._floor_added = True

    def _tick(self):
        try:
            tf = self.tf_buffer.lookup_transform(WORLD, CUBE_FRAME, rclpy.time.Time())
        except TransformException as ex:
            # waits until TF is available
            self.get_logger().debug(f"Waiting for TF {CUBE_FRAME}: {ex}")
            return

        pose = PoseStamped()
        pose.header.frame_id = WORLD
        pose.pose.position.x = tf.transform.translation.x
        pose.pose.position.y = tf.transform.translation.y
        pose.pose.position.z = tf.transform.translation.z
        pose.pose.orientation = tf.transform.rotation

        # upsert the cube as a collision box
        self.scene.add_box(OBJECT_NAME, pose, size=BOX_SIZE)

def main():
    rclpy.init()
    node = SceneSync()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: move_group.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <moveit/planning_interface/move_group_interface.h>

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("go_to_obj1");

  // ---- TF setup ----
  auto tf_buffer = std::make_shared<tf2_ros::Buffer>(node->get_clock());
  auto tf_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);

  const std::string base_frame = "base";     // or "base_link"
  const std::string obj_frame  = "obj_1";
  const std::string world_frame= "world";    // RViz fixed frame

  // Wait for TF to be available (up to 5s)
  const rclcpp::Duration timeout = rclcpp::Duration::from_seconds(5.0);
  while (rclcpp::ok() &&
         !tf_buffer->canTransform(base_frame, obj_frame, tf2::TimePointZero, tf2::durationFromSec(0.1))) {
    rclcpp::spin_some(node);
  }

  geometry_msgs::msg::TransformStamped t_bo;
  try {
    t_bo = tf_buffer->lookupTransform(base_frame, obj_frame, tf2::TimePointZero, timeout.to_chrono<std::chrono::seconds>());
  } catch (const tf2::TransformException& ex) {
    RCLCPP_ERROR(node->get_logger(), "TF lookup failed: %s", ex.what());
    rclcpp::shutdown();
    return 1;
  }

  // ---- Build target pose (add small Z approach) ----
  geometry_msgs::msg::PoseStamped target;
  target.header.frame_id = base_frame;
  target.header.stamp = node->now();
  target.pose.position.x = t_bo.transform.translation.x;
  target.pose.position.y = t_bo.transform.translation.y;
  target.pose.position.z = t_bo.transform.translation.z + 0.05;  // 5 cm above

  // Use obj_1 orientation directly (adjust to your gripper as needed)
  target.pose.orientation = t_bo.transform.rotation;

  // ---- MoveIt setup ----
  rclcpp::executors::SingleThreadedExecutor exec;
  exec.add_node(node);

  // Name must match your MoveIt group (e.g., "arm")
  moveit::planning_interface::MoveGroupInterface move_group(node, "arm");
  move_group.setPoseReferenceFrame(base_frame);
  move_group.setPlanningTime(5.0);

  move_group.setPoseTarget(target);
  moveit::planning_interface::MoveGroupInterface::Plan plan;
  bool ok = (move_group.plan(plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
  if (!ok) {
    RCLCPP_ERROR(node->get_logger(), "Planning failed.");
    rclcpp::shutdown();
    return 2;
  }

  auto res = move_group.execute(plan);
  if (res != moveit::planning_interface::MoveItErrorCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "Execution failed.");
  } else {
    RCLCPP_INFO(node->get_logger(), "Executed plan to obj_1.");
  }

  rclcpp::shutdown();
  return 0;
}
